### PR TITLE
Add File::with_options

### PR DESCRIPTION
**Pull request by Timmmm**
This provides a more fluent API to create files with options, and also avoids the need to import OpenOptions.

This implements @aldanor's [suggestion](https://github.com/rust-lang/rfcs/pull/2615#issuecomment-448591304) which was popular.

### CODE REVIEW

File: src/libstd/fs.rs [hunk 1]
```diff
@@ -397,6 +411,36 @@ impl File {
...
+    ///
+    /// ```no_run
+    /// use std::fs::File;
+    ///
+    /// fn main() -> std::io::Result<()> {
+    ///     let mut f = File::with_options().read(true).open("foo.txt")?;
+    ///     Ok(())
+    /// }
+    /// ```
+    #[stable(feature = "rust1", since = "1.0.0")]
```
Comment (jonas-schievink) [inline src/libstd/fs.rs]: This should start out as unstable, like all features do

Comment (Timmmm) [reply src/libstd/fs.rs]: Ah right. How do I do that? Just omit this line? (This is my first PR, sorry!)

Comment (jonas-schievink) [reply src/libstd/fs.rs]: You can check the guide (https://rust-lang.github.io/rustc-guide/stability.html) and also look at existing unstable features

File: src/libstd/fs.rs [hunk 2]
```diff
@@ -74,7 +74,7 @@ use crate::time::SystemTime;
 ///     Ok(())
 /// }
 /// ```
-///
+///
```
Comment (hellow554) [inline src/libstd/fs.rs]: please remove the trailing whitesapce

File: src/libstd/fs.rs [hunk 3]
```diff
@@ -397,6 +397,36 @@ impl File {
...
+    ///
+    /// ```no_run
+    /// use std::fs::File;
+    ///
+    /// fn main() -> std::io::Result<()> {
+    ///     let mut f = File::with_options().read(true).open("foo.txt")?;
+    ///     Ok(())
+    /// }
+    /// ```
+    #[unstable(feature = "with_options", issue = "55762")]
```
Comment (hellow554) [inline src/libstd/fs.rs]: #55762 is not really a tracking issue. Please have a look at "real" tracking issues: https://github.com/rust-lang/rust/labels/C-tracking-issue

Comment (Timmmm) [reply src/libstd/fs.rs]: Oh, do I have to create a tracking issue or can I just omit `, issue = "..."`?

**Comment by hellow554** [comment]
You do have to create one

File: src/libstd/fs.rs (hunk 3 above)
Comment (Timmmm) [reply src/libstd/fs.rs]: Never mind, I have created one.

**Comment by hellow554** [comment]
spurious error? @shepmaster could you trigger the build process again?

**Comment by lzutao** [comment]
That's docker hub problem: https://status.docker.com/pages/incident/533c6539221ae15e3f000031/5da5fc78c7b1c11cd8603721

**Comment by hellow554** [comment]
@Timmm are you able to fix that or do you need help? (hint: take a look at https://github.com/rust-lang/rust/blob/f54911c6f2971b98ede2d21ab8cabd7daa1110ba/src/libstd/fs.rs#L97-L105)

**Comment by Timmmm** [comment]
At last :-D Sorry for using the CI system as my build tool... Next time I will compile it myself!

**Comment by JohnCSimon** [comment]
Ping from triage: 

@shepmaster Can you please review this PR? Thanks
CC: @Timmmm

**Comment by JohnCSimon** [comment]
Pinging again from triage:

@shepmaster Can you please review this PR? Thanks
CC: @Timmmm @hellow554

**Comment by Mark-Simulacrum** [comment]
r? @Mark-Simulacrum 

@bors r+

Seems like some libs folks were in favor of this on the associated RFC so we can probably move ahead here (it's an unstable feature anyway).

File: src/libstd/fs.rs [hunk 4]
```diff
@@ -397,6 +399,37 @@ impl File {
         OpenOptions::new().write(true).create(true).truncate(true).open(path.as_ref())
     }
 
+    /// Returns a new OpenOptions object.
```
Comment (lzutao) [inline src/libstd/fs.rs]: add link for this
```suggestion
    /// Returns a new [OpenOptions] object.
```

File: src/libstd/fs.rs [hunk 5]
```diff
@@ -397,6 +399,37 @@ impl File {
...
+    ///
+    /// This function returns a new OpenOptions object that you can use to
+    /// open or create a file with specific options if `open()` or `create()`
+    /// are not appropriate.
+    ///
+    /// It is equivalent to `OpenOptions::new()` but allows you to write more
```
Comment (lzutao) [inline src/libstd/fs.rs]:
```suggestion
    /// It is equivalent to [`OpenOptions::new()`] but allows you to write more
```

File: src/libstd/fs.rs [hunk 6]
```diff
@@ -397,6 +399,37 @@ impl File {
...
+    /// readable code. Instead of `OpenOptions::new().read(true).open("foo.txt")`
+    /// you can write `File::with_options().read(true).open("foo.txt"). This
+    /// also avoids the need to import `OpenOptions`.
+    ///
+    /// See the [`OpenOptions::new`] function for more details.
```
Comment (lzutao) [inline src/libstd/fs.rs]:
```suggestion
```

File: src/libstd/fs.rs [hunk 7]
```diff
@@ -397,6 +399,37 @@ impl File {
         OpenOptions::new().write(true).create(true).truncate(true).open(path.as_ref())
     }
 
+    /// Returns a new OpenOptions object.
+    ///
+    /// This function returns a new OpenOptions object that you can use to
+    /// open or create a file with specific options if `open()` or `create()`
+    /// are not appropriate.
```
Comment (lzutao) [inline src/libstd/fs.rs]: Remove theses lines.

**Pull request merged**